File: divedave-ios/divedave Shared/Components/Menu/InfoPanel.swift
import SceneKit
import SpriteKit

@MainActor
final class InfoPanel {
    private var baseDepth: CGFloat
    private var panel: SKSpriteNode
    private var daveImage: AnimatedSprite
    private var score1: SKSpriteNode
    private var score2: SKSpriteNode
    private var score3: SKSpriteNode
    private var tryAgain: SKLabelNode
    private var scene: SKScene
    private let scoreSpacing: CGFloat

    init(scene: SKScene, depth: CGFloat) {
        self.scene = scene
        self.baseDepth = depth
        self.scoreSpacing = 275

        panel = SKSpriteNode(imageNamed: "panel")
        panel.position = CGPoint(x: GameState.shared.metrics.width / 2, y: GameState.shared.metrics.height / 2)
        panel.zPosition = baseDepth
        panel.isHidden = true
        scene.addChild(panel)

        daveImage = AnimatedSprite(spritesheetName: "divedave-emotions", frameWidth: 150, frameHeight: 160)
        daveImage.position = CGPoint(x: GameState.shared.metrics.width / 2, y: GameState.shared.metrics.height / 2 + 175)
        daveImage.setScale(2)
        daveImage.zPosition = baseDepth + 1
        daveImage.isHidden = true
        scene.addChild(daveImage)

        score1 = SKSpriteNode(imageNamed: "sign")
        score1.position = CGPoint(x: GameState.shared.metrics.width / 2 - scoreSpacing, y: GameState.shared.metrics.height / 2 - 325)
        score1.zPosition = baseDepth + 2
        score1.isHidden = true
        scene.addChild(score1)

        score2 = SKSpriteNode(imageNamed: "sign")
        score2.position = CGPoint(x: GameState.shared.metrics.width / 2, y: GameState.shared.metrics.height / 2 - 325)
        score2.zPosition = baseDepth + 2
        score2.isHidden = true
        scene.addChild(score2)

        score3 = SKSpriteNode(imageNamed: "sign")
        score3.position = CGPoint(x: GameState.shared.metrics.width / 2 + scoreSpacing, y: GameState.shared.metrics.height / 2 - 325)
        score3.zPosition = baseDepth + 2
        score3.isHidden = true
        scene.addChild(score3)

        tryAgain = SKLabelNode(fontNamed: "DrawvidHand-Regular")
        tryAgain.text = "tap to dive again"
        tryAgain.fontSize = 65
        tryAgain.position = CGPoint(x: GameState.shared.metrics.width / 2, y: 75)
        tryAgain.zPosition = baseDepth + 2
        tryAgain.isHidden = true
        scene.addChild(tryAgain)
    }

    func display(result: String, strings: [String], frame: Int, scores: [Int]?) {
        var height = GameState.shared.metrics.height / 2 + 5
        let resultColor: SKColor = (result == "FAILED DIVE" || result == "GAME OVER") ? Game.customRed : Game.customGreen
        tryAgain.fontColor = resultColor
        let resultLabel = SKLabelNode(fontNamed: "DrawvidHand-Regular")
        resultLabel.text = result
        resultLabel.fontSize = 80
        resultLabel.fontColor = resultColor
        resultLabel.position = CGPoint(x: GameState.shared.metrics.width / 2, y: height + 350)
        resultLabel.zPosition = baseDepth + 1
        scene.addChild(resultLabel)

        let lineSpacing: CGFloat = 75
        for string in strings {
            let color = (string.contains("rotations") && (result == "FAILED DIVE" || result == "GAME OVER")) ? SKColor.red : SKColor.black
            let stringLabel = SKLabelNode(fontNamed: "DrawvidHand-Regular")
            stringLabel.text = string
            stringLabel.fontSize = 80
            stringLabel.fontColor = color
            stringLabel.position = CGPoint(x: GameState.shared.metrics.width / 2, y: height)
            stringLabel.zPosition = baseDepth + 1
            scene.addChild(stringLabel)
            height -= lineSpacing
        }

        var scoreLabels: [SKLabelNode] = []
        if let scores = scores {
            var width = GameState.shared.metrics.width / 2 - scoreSpacing
            for score in scores {
                let scoreLabel = SKLabelNode(fontNamed: "DrawvidHand-Regular")
                scoreLabel.text = "\(score)"
                scoreLabel.fontSize = 100
                scoreLabel.position = CGPoint(x: width, y: GameState.shared.metrics.height / 2 - 350)
                scoreLabel.zPosition = baseDepth + 3
                scoreLabel.fontColor = .red
                scoreLabel.alpha = 0
                scoreLabel.setScale(0.3)
                scene.addChild(scoreLabel)
                scoreLabels.append(scoreLabel)
                width += scoreSpacing
            }
        }

        panel.isHidden = false
        if (frame >= 0) {
            daveImage.texture = daveImage.frames[frame]
        } else {
            daveImage.texture = daveImage.frames[0]
        }
        daveImage.isHidden = false

        tryAgain.position = CGPoint(x: GameState.shared.metrics.width / 2, y: 100)

        if scores != nil {
            let signs = [score1, score2, score3]
            let signBaseScale = score1.xScale

            for sign in signs {
                sign.alpha = 0
                sign.setScale(signBaseScale * 0.3)
                sign.isHidden = false
            }
            tryAgain.alpha = 0
            tryAgain.isHidden = true

            var seq: [SKAction] = []
            for idx in 0..<min(signs.count, scoreLabels.count) {
                let sign = signs[idx]
                let label = scoreLabels[idx]
                seq.append(SKAction.wait(forDuration: 0.4))
                seq.append(SKAction.run {
                    sign.run(SKAction.group([
                        SKAction.fadeIn(withDuration: 0.1),
                        SKAction.sequence([
                            SKAction.scale(to: signBaseScale * 1.2, duration: 0.12),
                            SKAction.scale(to: signBaseScale, duration: 0.08)
                        ])
                    ]))
                    label.run(SKAction.group([
                        SKAction.fadeIn(withDuration: 0.1),
                        SKAction.sequence([
                            SKAction.scale(to: 1.2, duration: 0.12),
                            SKAction.scale(to: 1.0, duration: 0.08)
                        ])
                    ]))
                })
            }
            seq.append(SKAction.wait(forDuration: 0.4))
            seq.append(SKAction.run { [weak self] in
                guard let self = self else { return }
                self.tryAgain.isHidden = false
                let blink = SKAction.repeatForever(SKAction.sequence([
                    SKAction.fadeAlpha(to: 0.2, duration: 0.5),
                    SKAction.fadeAlpha(to: 1.0, duration: 0.5)
                ]))
                self.tryAgain.run(SKAction.sequence([
                    SKAction.fadeIn(withDuration: 0.2),
                    blink
                ]), withKey: "blink")
            })
            scene.run(SKAction.sequence(seq))
        } else {
            score1.isHidden = true
            score2.isHidden = true
            score3.isHidden = true
            tryAgain.isHidden = true
        }
    }

    func close() {
        tryAgain.removeAction(forKey: "blink")
        tryAgain.isHidden = true
        panel.isHidden = true
        daveImage.isHidden = true
        score1.isHidden = true
        score2.isHidden = true
        score3.isHidden = true
    }
}

File: divedave-web/src/components/menu/InfoPanel.js
import { HEIGHT, WIDTH } from "../../util/Constants.js";
import { Haptics } from "../../util/Haptics.js";
import { IS_MOBILE } from "../../util/Utilities.js";

const REVEAL_BEAT_MS = 400;
const REVEAL_TWEEN_MS = 200;

export class InfoPanel extends Phaser.GameObjects.Group {
  constructor(scene, depth) {
    super(scene);
    this.baseDepth = depth;
    this.panel = scene.add
      .image(WIDTH / 2, HEIGHT / 2, "panel")
      .setVisible(false)
      .setScrollFactor(0)
      .setDepth(this.baseDepth);
    this.daveimage = scene.add
      .sprite(WIDTH / 2, HEIGHT / 2 - 175, "davemotions")
      .setVisible(false)
      .setScrollFactor(0)
      .setScale(2)
      .setFrame(2)
      .setDepth(this.baseDepth + 1);
    this.score1 = scene.add
      .image(WIDTH / 2 - 275, HEIGHT / 2 + 325, "sign")
      .setVisible(false)
      .setScrollFactor(0)
      .setDepth(this.baseDepth + 2);
    this.score2 = scene.add
      .image(WIDTH / 2, HEIGHT / 2 + 325, "sign")
      .setVisible(false)
      .setScrollFactor(0)
      .setDepth(this.baseDepth + 2);
    this.score3 = scene.add
      .image(WIDTH / 2 + 275, HEIGHT / 2 + 325, "sign")
      .setVisible(false)
      .setScrollFactor(0)
      .setDepth(this.baseDepth + 2);
    this.tryAgain = scene.add
      .bitmapText(
        125,
        75,
        "red-arial",
        (IS_MOBILE ? "tap" : "click") + " to dive again",
        65
      )
      .setOrigin(0.5)
      .setDepth(this.baseDepth + 2)
      .setScrollFactor(0)
      .setVisible(false);
  }

  display(scene, result, strings, frame, scores, sceneHeight) {
    let height = HEIGHT / 2 + 5;
    scene.add
      .bitmapText(
        WIDTH / 2,
        height - 375,
        result === "FAILED DIVE" || result === "GAME OVER"
          ? "red-arial"
          : "green-arial",
        result,
        80
      )
      .setOrigin(0.5)
      .setDepth(this.baseDepth + 1)
      .setScrollFactor(0);
    strings.forEach((string) => {
      scene.add
        .bitmapText(
          WIDTH / 2,
          height,
          string.includes("rotations") &&
            (result === "FAILED DIVE" || result === "GAME OVER")
            ? "red-arial"
            : "black-arial",
          string,
          80
        )
        .setOrigin(0.5)
        .setDepth(this.baseDepth + 1)
        .setScrollFactor(0);
      height += 75;
    });

    this.panel.setVisible(true);
    this.daveimage.setFrame(frame);
    this.daveimage.setVisible(true);
    this.tryAgain.setPosition(WIDTH / 2, sceneHeight - 100);

    if (!scores) {
      this.score1.setVisible(false);
      this.score2.setVisible(false);
      this.score3.setVisible(false);
      this.tryAgain.setVisible(false);
      return;
    }

    const signs = [this.score1, this.score2, this.score3];
    const numberLabels = scores.map((score, i) =>
      scene.add
        .bitmapText(
          WIDTH / 2 - 275 + i * 275,
          HEIGHT / 2 + 325,
          "red-arial",
          score,
          100
        )
        .setOrigin(0.5)
        .setDepth(this.baseDepth + 3)
        .setScrollFactor(0)
        .setAlpha(0)
        .setScale(0.3)
    );

    signs.forEach((sign) => {
      sign.setVisible(true).setAlpha(0).setScale(0.3);
    });
    this.tryAgain.setVisible(true).setAlpha(0);

    const isStillValid = (obj) => obj && obj.active;

    scores.forEach((_score, idx) => {
      scene.time.delayedCall(REVEAL_BEAT_MS * (idx + 1), () => {
        if (!isStillValid(signs[idx]) || !isStillValid(numberLabels[idx])) {
          return;
        }
        Haptics.impactMedium();
        scene.tweens.add({
          targets: [signs[idx], numberLabels[idx]],
          alpha: 1,
          scaleX: { from: 0.3, to: 1.0 },
          scaleY: { from: 0.3, to: 1.0 },
          ease: "Back.easeOut",
          duration: REVEAL_TWEEN_MS,
        });
      });
    });

    scene.time.delayedCall(REVEAL_BEAT_MS * (scores.length + 1), () => {
      if (!isStillValid(this.tryAgain)) return;
      scene.tweens.add({
        targets: this.tryAgain,
        alpha: 1,
        duration: REVEAL_TWEEN_MS,
      });
    });
  }

  close() {
    this.tryAgain.setVisible(false);
    this.panel.setVisible(false);
    this.daveimage.setVisible(false);
    this.score1.setVisible(false);
    this.score2.setVisible(false);
    this.score3.setVisible(false);
  }
}
